### PR TITLE
First implementation of test tier tagging support

### PR DESCRIPTION
**Pull request by Jurisak**
Added resultDB query plugin. It is not functional now.
We need to wait till resultdb listener update.

### CODE REVIEW

**Comment by gsr-shanks** [comment]
Please add docstrings.

**Comment by Jurisak** [comment]
I'll add them and also with tests. @gsr-shanks did you enable travis for this project?

**Comment by gsr-shanks** [comment]
No, I did not enable travis.

**Comment by gsr-shanks** [comment]
Please add docs as well.

File: plugins/morph_resultsdb.py
Comment (gsr-shanks) [inline plugins/morph_resultsdb.py]: Can we retry this for 3 times with sleep of 120 seconds before raising ResultsDBApiException ?

Comment (Jurisak) [reply plugins/morph_resultsdb.py]: Sure, I'll do.

**Comment by gsr-shanks** [comment]
Nice. Getting NVR from ci-message should be the default. Thanks.

**Comment by gsr-shanks** [comment]
`creating /usr/lib/python2.7/site-packages/plugins
copying build/lib/plugins/morph_messagehub.py -> /usr/lib/python2.7/site-packages/plugins
copying build/lib/plugins/morph_resultsdb.py -> /usr/lib/python2.7/site-packages/plugins
creating /usr/lib/python2.7/site-packages/plugins/library
copying build/lib/plugins/library/messagehub.py -> /usr/lib/python2.7/site-packages/plugins/library
copying build/lib/plugins/__init__.py -> /usr/lib/python2.7/site-packages/plugins
creating /usr/lib/python2.7/site-packages/lib
copying build/lib/lib/logging_conf.py -> /usr/lib/python2.7/site-packages/lib
copying build/lib/lib/__init__.py -> /usr/lib/python2.7/site-packages/lib`

packages are getting installed in site-packages basedir instead of site-packages/metamorph/

**Comment by Jurisak** [comment]
Fixed in #6. Please merge that pull request as soon as possible.

**Comment by gsr-shanks** [comment]
_plugins/morph_resultsdb.py_ should be _metamorph/plugins/morph_resultsdb.py_

**Comment by Jurisak** [comment]
@gsr-shanks Sure i know. I've updated #6 pull request. Please merge it

File: plugins/morph_resultsdb.py
Comment (gsr-shanks) [inline plugins/morph_resultsdb.py]: should be `from metamorph.lib.logging_conf`

Comment (Jurisak) [reply plugins/morph_resultsdb.py]: Yes but also no. In current state no metamoprh folder exists so it's correct. I'll solve it in a minute.

**Comment by Jurisak** [comment]
I would like to also mention that this plugin is not working yet. You can only try it on resultsdb devel stage because resultsdb listener changes were not moved into production, yet.

**Comment by gsr-shanks** [comment]
Ok, could you please email me the command that I can try with resultsdb devel stage ?

**Comment by gsr-shanks** [comment]
@robnester-rh could you please take a look

**Comment by Jurisak** [comment]
@gsr-shanks I'll create ansible module after resultsdb updater changes will be merged in production.

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: You could consider reworking this method to use the requests.raise_for_status() method to catch issues with communicating with the API.
The following seems to be a bit more compact to achieve the same result.

```
    logging.debug('Running resultsdb API query with this url: {0} and options {1}'.format(url, url_options))
    try:
        response = requests.get(url, params=url_options)
        response.raise_for_status()
        return response.json()
    except requests.HTTPError as e:
        if attempt < 3:
            attempt += 1
            time.sleep(120)
            self.query_resultsdb(url, url_options, attempt)
        else:
            logging.error("ERROR: Unable to access resultsdb site.")
            logging.error("ERROR: {0}".format(e.args)) # If you want to pass the exact error.
```

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: From your docstring I think you're wanting to create a dictionary of the following structure:
```
formatted_output['job_names']<key = some job name> = <data from resultsdb_data>
```

If this is the case, then this method should be updated as such:

```
formatted_output = dict(job_names=dict())
for single_job in resultsdb_data:
    formatted_output['job_names'][single_job] = resultsdb_data[single_job] # if resultsdb_data is a dictionary of data, which I'm assuming it is from your existing structure.
return formatted_output
```

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: nit: correct to "Formats"

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: nit: correct typo  from 'resultsbd' to 'resultsdb'

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py:24]: I don't really think it's necessary to define MINUTE as a class constant. This seems to lead to having to do math on its value instead of just passing in given integers. Is it really any more clear to say sleep(self.MINUTE + self.MINUTE) instead of saying sleep(120) ?

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: See above comment regarding defining this as a class constant.

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: The get_test_tier_metadata method appears to be an unused class method, is this for some future usage?

I'm curious also in that its name seems to suggest that it would return test tier metadata, but it's calling get_resultsdb_data, which doesn't return test tier metadata explicitly it seems?

File: metamorph/plugins/morph_resultsdb.py
Comment (robnester-rh) [inline metamorph/plugins/morph_resultsdb.py]: I don't know if it's a RH standard, but I generally separate stdlib imports, 3rd party imports and then project local imports by a blank line, such as the following:
```
import argparse
import logging
import logging.config
import json
import time
import os

import requests

from metamorph.lib.logging_conf import setup_logging
```

**Review by robnester-rh — CHANGES_REQUESTED** [review]
I think you have one "need to fix" in morph_resultsdb.py around line 100 regarding creating your dictionary of data that you're returning in the setup_output_data() method

Additionally, I would encourage looking at the comment regarding usage of .raise_for_status around line 47 in  the query_resultsdb() method in the same file.

Otherwise, a few nits / typos to fix.

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: Thanks that's a good attitude. I'll fix it.

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py:24]: I generally don't like having magic numbers in my code.
@gsr-shanks  What do you think?

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: Good catch. Yes I want to use it in the main function.
I'll think about a new name.

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: Thanks

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: We can discuss this in above comment.

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: Wow this looks really good. I didn't know about raise_for_status() method.
Thanks for this suggestion.

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: Good catch. Thanks

File: metamorph/plugins/morph_resultsdb.py
Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: resultsdb_data is list of queried resultsdb data (dictionaries). Therefore it can't be that way.

Comment (robnester-rh) [reply metamorph/plugins/morph_resultsdb.py]: If resultsdb_data is a list of dictionaries, it still wouldn't work the way you've written it here I think.
If I understand you correctly, the following should be analogous to what you'd have:

```
resultsdb_data = [ {'job_names': 'a'}, {'job_names': 'b'}, {'job_names': 'c'}, {'job_names': 'd'} ]
formatted_output = {}
for single_job in resultsdb_data:
    formatted_output[single_job['job_names']] += single_job
print(formatted_output)
```
And this doesn't work. 

Can you clarify what your formatted_output dictionary structure should look like?

Comment (Jurisak) [reply metamorph/plugins/morph_resultsdb.py]: Yes I know how you meant it but resultsdb_data looks like this:
`resultsdb_data [ { "data": { .. }, "job_names": "...", "id": [], "outcome": "PASS" }, { "data": ..., "job_names": ... }]`

For output dictionary structure see: [https://url.corp.redhat.com/ed75ec6]

Comment (robnester-rh) [reply metamorph/plugins/morph_resultsdb.py]: Ah, ok. I see now. I wasn't assuming duplicate dict structures. Thanks for the clarification!

File: metamorph/plugins/morph_resultsdb.py
Comment (gsr-shanks) [reply metamorph/plugins/morph_resultsdb.py]: I don't think we need to do math in this case. sleep(self.MINUTE) is good enough time to retry if the first try is unsuccessful.

File: metamorph/plugins/morph_resultsdb.py
Comment (gsr-shanks) [reply metamorph/plugins/morph_resultsdb.py:24]: @Jurisak commented below.

**Comment by gsr-shanks** [comment]
There is still one change requested in this PR.

**Comment by gsr-shanks** [comment]
@robnester-rh Looks like the changes requested are all in. Could you please merge this if all good ?

File: tests/sources/resultsdb_output.json
Comment (robnester-rh) [inline tests/sources/resultsdb_output.json]: fix typo "someursl"

File: tests/sources/resultsdb_output_result.json
Comment (robnester-rh) [inline tests/sources/resultsdb_output_result.json]: fix typo "someursl"

**Review by robnester-rh — COMMENTED** [review]
fix typo "someursl" where mentioned.

**Comment by Jurisak** [comment]
@robnester-rh Updated

**Review by robnester-rh — APPROVED** [review]

**Pull request merged**
